packages/realm-react/src/RealmProvider.tsx

```diff
@@ -41,43 +41,73 @@ type ProviderProps = PartialRealmConfiguration & {
    * instance outside of a component that uses the Realm hooks.
    */
   realmRef?: React.MutableRefObject<Realm | null>;
+  /**
+   * The fallback component to render if the Realm is not open.
+   */
+  fallback?: React.ComponentType<unknown> | React.ReactElement | null | undefined;
   children: React.ReactNode;
-};
+} & PartialRealmConfiguration;
+
+/**
+  Represents the provider returned from using an existing realm at context creation i.e. `createRealmContext(new Realm))`.
+ */
+export type RealmProviderFromRealmInstanceFC = React.FC<Pick<RealmProviderProps, "children">>;
+
+/**
+ * Represents the provider returned from using a Realm configuration at context creation i.e. `createRealmContext({schema: []}))`.
+ */
+export type RealmProviderFromConfigFC = React.FC<
+  Pick<RealmProviderProps, "closeOnUnmount" | "realmRef" | "fallback" | "children" | keyof PartialRealmConfiguration>
+>;
+
+/**
+ * Explicitly sets the unpicked properties of a type to never instead of dropping them like in Pick.
+ * Useful for ensuring different prop types are mutually exclusive as React expects the union type
+ * of different prop types to include all the fields.
+ */
+type RestrictivePick<T, K extends keyof T> = Pick<T, K> & { [RestrictedKey in keyof Omit<T, K>]?: never };
+
+/**
+ * Represents properties of a generic RealmProvider where the realm is set and therefore all other
+ * props should be disallowed.
+ */
+export type RealmProviderWithRealmInstanceProps = RestrictivePick<RealmProviderProps, "realm" | "children">;
+
+/*
+ * Represents properties of a generic RealmProvider where Realm configuration props are used and no realm is set
+ */
+export type RealmProviderWithConfigurationProps = RestrictivePick<
+  RealmProviderProps,
+  "closeOnUnmount" | "realmRef" | "fallback" | "children" | keyof PartialRealmConfiguration
+>;
+
+/**
+ * Represents the provider returned from creating context with no arguments (including the default context).
+ * Supports either passing a `realm` as a property or the schema configuration.
+ */
+export type GeneralizedRealmProviderFC = React.FC<
+  RealmProviderWithRealmInstanceProps | RealmProviderWithConfigurationProps
+>;
```

Comment:
This is definetely simpler, I like the new names but with some there's some complications:
1.  It's a very subtle distinction but `RealmProviderFromRealmProps != RealmProviderWithRealmProps` .

If the provider was created from `createRealmContext(realm)` then there's no `realm` to pass as props again (unless we want to support overriding it after creation but I think it'll lead to people passing it twice a lot, it'd make sense for them to use the DynamicRealmProvider then).
This is not the case for `RealmProviderFromConfigurationProps` since you can basically override all the options you pass to `createRealmContext`, so `RealmProviderFromConfigurationProps === RealmProviderWithConfigurationProps`

I also personally like the existence of a singular `RealmProviderProps` and doing the picking from there because it gives us a one point of entry for all the props a user sees when they get a `DynamicRealmProvider` and then we can create restrictive subsets for different scenarios.

```ts
export type DynamicRealmProvider = React.FC<RealmProviderFromConfigurationProps | RealmProviderFromRealmProps>;
```


2. This `DynamicRealmProvider` will allow you to set both `realm` **and** configuration props at the same time. We can throw an error in this case ourselves on runtime but I think it's nicer UX to enforce it through types (unless it can seem like maintenece cost is too high... I think it's worth it). I don't think TS it smart enough to understand the union and separate it when enforcing props, which is why I use `RestrictedPick`, it allows for mutually exclusive props to be used as a union.

Reply:
Yeah I think if we could prevent having to think too much about `From`/`With` distinction in the types that'd be good as well 🙂 So I guess there are 4 allowed combinations for props then:
* With configuration
* With realm
* Just empty (incl. children)
* With configuration *or* realm (dynamic)

E.g.:
```ts
type RealmProviderWithConfigurationProps = {
  closeOnUnmount?: boolean;
  realmRef?: React.MutableRefObject<Realm | null>;
  fallback?: React.ComponentType<unknown> | React.ReactElement | null | undefined;
  children: React.ReactNode;
} & PartialRealmConfiguration;

type RealmProviderWithRealmProps = {
  realm: Realm; // Note no `?`
  children: React.ReactNode;
};

type RealmProviderWithEmptyProps = {
  children: React.ReactNode;
};

type DynamicRealmProviderProps = { /* TODO (further below) */ };
```

This gives us 4 provider types (avoiding to mix `From` and `With` in these types used):

```ts
export type RealmProviderWithConfiguration = React.FC<RealmProviderWithConfigurationProps>;
export type RealmProviderWithRealm = React.FC<RealmProviderWithRealmProps>;
export type RealmProviderWithEmpty = React.FC<RealmProviderWithEmptyProps>;
export type DynamicRealmProvider = React.FC<DynamicRealmProviderProps>;

export type RealmProvider = RealmProviderWithConfiguration | RealmProviderWithEmpty | DynamicRealmProvider;
```

To address the `DynamicRealmProvider`, as you pointed we should have a type that:
* Allows all keys of `RealmProviderWithRealmProps`, while disallowing (`never | undefined`) the excess or non-shared keys of `RealmProviderWithConfigurationProps`.
* Or, allows all keys of `RealmProviderWithConfigurationProps`, while disallowing the excess or non-shared keys of `RealmProviderWithRealmProps`.

In order to not have to maintain adding/modifying individual keys in the types I believe this should work:

```ts
/**
 * A type with all properties of `T`, as well as any excess
 * properties on `Excess` set to `never | undefined`.
 */
type WithExcessAsNever<T, Excess> = T & { [K in keyof Omit<Excess, keyof T>]?: never };

type DynamicRealmProviderProps =
  | WithExcessAsNever<RealmProviderWithRealmProps, RealmProviderWithConfigurationProps>
  | WithExcessAsNever<RealmProviderWithConfigurationProps, RealmProviderWithRealmProps>;
```

Perhaps ☝️ could even be simplified further.

The `create` functions are similar to before (haven't renamed the `From` in these names as its *from* the argument passed):
```ts
export function createRealmProviderFromRealm(
  // Note: Was previously `Realm | null` but seems like the callers always pass a realm here.
  realm: Realm,
  ...
): RealmProviderWithEmpty { ... }
export function createRealmProviderFromConfiguration(...): RealmProviderWithConfiguration { ... }
export function createDynamicRealmProvider(): DynamicRealmProvider { ... }
export function createRealmProvider(...): RealmProvider { ... }
```